components/brave_wallet/browser/json_rpc_service.cc

```diff
@@ -1395,6 +1396,190 @@ void JsonRpcService::ContinueGetERC721TokenBalance(
                           mojom::ProviderError::kSuccess, "");
 }
 
+void JsonRpcService::GetERC721Metadata(const std::string& contract_address,
+                                       const std::string& token_id,
+                                       const std::string& chain_id,
+                                       GetERC721MetadataCallback callback) {
+  auto network_url = GetNetworkURL(prefs_, chain_id, mojom::CoinType::ETH);
+  if (!network_url.is_valid()) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  if (!EthAddress::IsValidAddress(contract_address)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  uint256_t token_id_uint = 0;
+  if (!HexValueToUint256(token_id, &token_id_uint)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  std::string function_signature;
+  if (!erc721::TokenUri(token_id_uint, &function_signature)) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
+    return;
+  }
+
+  auto internal_callback =
+      base::BindOnce(&JsonRpcService::OnGetSupportsInterfaceERC721Metadata,
+                     weak_ptr_factory_.GetWeakPtr(), contract_address,
+                     function_signature, network_url, std::move(callback));
+
+  GetSupportsInterface(contract_address, kERC721MetadataInterfaceId,
+                       std::move(internal_callback));
+}
+
+void JsonRpcService::OnGetSupportsInterfaceERC721Metadata(
+    const std::string& contract_address,
+    const std::string& function_signature,
+    const GURL& network_url,
+    GetERC721MetadataCallback callback,
+    bool is_supported,
+    mojom::ProviderError error,
+    const std::string& error_message) {
+  if (error != mojom::ProviderError::kSuccess) {
+    std::move(callback).Run("", error, error_message);
+    return;
+  }
+
+  if (!is_supported) {
+    std::move(callback).Run(
+        "", mojom::ProviderError::kInvalidParams,
+        l10n_util::GetStringUTF8(IDS_WALLET_INVALID_PARAMETERS));
```

Comment:
why would we use invalid params if this isn't supported? maybe makes more sense for kMethodNotSupported?

Reply:
I guess the case for invalid params is that the client supplied parameters for which we cannot obtain a result.. It's kind of a stretch.  I've mentioned it elsewhere, but I wasn't sure if kMethodNotSupported is applicable because I'm not sure if we could say the RPC method itself is not supported

Reply:
Sounds good to me.  Let's go with `kMethodNotSupported` for now like the other one.

There may be an argument to more cleanly divide Brave Wallet APIs from the eth JSON RPC APIs.  I'm thinking something along the lines of treating the JsonRpcService as a lower level primitive that's used by dapps and a different higher level Brave specific service for things like metadata fetching.  The higher level service could have it's own set of errors, and we could avoid any conflation our specific errors with eth JSON RPC protocol errors.

Reply:
Updated to use kMethodNotSupported